### ISSUE TITLE
Support for Multiple Domains

### ISSUE THREAD

**Issue by dariopedol** [kyma-project/kyma#5578]
**Description**

We would like to have multiple domains pointing to our cluster. E.g.:

*.test.com
*.test2.com

Would be pointing to the same cluster. A simple CNAME is not possible due to the certificates. 

**Reasons**

Kyma provides OOTB namespaces for different environments (stage/production). We need to be able to deploy API definitions that include the TLD to have production URLs.

**Comment by salvocuccurullo** [comment]
I would suggest to do not close this ticket because it is quite important.
I'm managing a Kyma cluster in production and having an alternative CNAME for accessing it would be very useful. The reason we need it, on our case, is the following:

Let's pretend we have the latest version of Kyma running on kyma01.mydomain.com.
Then a new Kyma release is out and we don't want to execute an in-place upgrade on a production cluster but we want to spawn a new one with the latest Kyma version, migrate the installed applications and finally just switch the CNAME to the new cluster.
This could also help us to manage our external vendors, avoiding to inform them to switch on their side the kyma url every time we move the installation to a new cluster.

The TLS certificate is not a big issue because it is possible to use ALT NAMES during certificate creation:

Example:
`
docker run -it --name certbot --rm  \
    -v "$(pwd)/letsencrypt:/etc/letsencrypt" \
    certbot/certbot \
    certonly \
    -m <EMAIL> --agree-tos --no-eff-email \
    --manual \
    --manual-public-ip-logging-ok \
    --preferred-challenges dns \
    --server https://acme-v02.api.letsencrypt.org/directory \
    -d "*.kyma01.cluster.mydomain1.com, *.kyma-prod.mydomain2.com"
`
Could you please at least evaluate a potential solution?
Thanks!

**Comment by piotrmsc** [comment]
So to recap, you would like to have kyma ingress listening on multiple domains, right?
Example that comes to my mind is : under domainA you would like to serve kyma internally and under domainB serve it publicly.

This is a big topic as domain is one of the parameters passed during installation to kyma components like ingress, application connector, cert jobs etc. Another example is our micro API Gateway exposing service on kyma domain. With this new feature, depending on number of configured domains it should support multiple domains as well. It's possible but for sure will be a big topic. 

Let's first make sure we understand fully each other and then I will try to complie plan for it.
Would be nice if you could join our open slack and we could talk there, even make a call if necessary ;)

When it comes to @salvocuccurullo comment, this flow should work for you OOTB. What exactly do you need in kyma in such flow? You'd like to have cert issuance automated?

@salvocuccurullo @dariopedol

**Comment by dariopedol** [comment]
You're right about what you presume and I also see the problems. My use case would be to serve, for example, stage namespace to one domain and prod namespace to another. At the moment this is not possible as api definitions are not namespace agnostic (duplicate domains lead to error).
We solved this by just setting up multiple clusters which serves us well for now. 
Regarding the API Gateway - I would add a new parameter to the api definition, by default serving all domains.

**Comment by salvocuccurullo** [comment]
Hello @dariopedol, @piotrmsc 
What we really need is the following real life example:
- We have Kyma 1.7 running and we don't want to apply an in-place upgrade to 1.8
- We spawn a new cluster and we install Kyma 1.8
- If the installation is successful we switch the CNAME to the new domain

`                        
                                              | ---> *.kyma1-7.mydomain.com 
*.kyma.mydomain.com ----|
                                              |    *.kyma1-8.mydomain.com
`

**Comment by DavidZisky** [comment]
No update on this? It's a pretty important feature especially for enterprises

**Comment by piotrmsc** [comment]
Well, we have 2 different requests here. One is to expose APIs separated by namespaces -> host used to expose API created in namespace A contains this name as a port of it. Second, allow access to kyma cluster under to different hosts which requires DNS management with updating CNAME like mentioned by @salvocuccurullo. When it comes to the first one, we have in our mind but currently, we are focused on different topics. When it comes to the second one, we do not offer DNS management and this is something that should be done by the admins that issue production TLS certificates and manage DNS entries.

But still, I totally understand use case from @salvocuccurullo but don't see how kyma could do that, this is work to be done on infra level.

**Comment by DavidZisky** [comment]
@piotrmsc I was referring to @salvocuccurullo request. And I am happy to explain a bit more because that's something I am trying to do since weeks and Kyma is not helping. So first of all - Kyma does not need to do DNS management - not at all, I totally agree with you that it's infra team responsibility. We only need Kyma to support being behind the reverse proxy. Currently, it doesn't because it does a bunch of redirects internally - if you take @salvocuccurullo example - assuming that *.kyma.mydomain.com is a VM with nginx installed which does redirect either to *.kyma1-7.mydomain.com or *.kyma1-8.mydomain.com - so when you open in the browser console.kyma.mydomain.com you will get redirected to console.kyma1-7.mydomain.com (by nginx) but then Kyma will redirect you to dex.kyma1-7.mydomain.com (and this redirect is not going via nginx so I can't rewrite it)

So what we are looking for is the ability to specify somehow additional "proxy" URL in Kyma. Or to simply put it in different words - I agree that work needs to mainly done on infra level but we need some support from Kyma here because I am simply not able to achieve that right now

**Comment by salvocuccurullo** [comment]
What @DavidZisky wrote is exactly what I meant. We can work on infra level for managing the multiple domains and setup either nginx as reverse proxy or a sort of gateway. The problem is that any internal Kyma redirection is not easily manageable.

**Comment by piotrmsc** [comment]
Still valid

**Comment by DavidZisky** [comment]
We made Kyma working behind proxy, it required few changes to some configmaps of Kyma components. I'll write later today all the details, maybe that'll help you

**Comment by piotrmsc** [comment]
Cool! I'm very open ! 🎉  We do not drop this one, it will be on our radar but in the near future. Currently, we focus on finishing different topic :) 
Guys, maybe you can join our open source slack and we can discuss, even make a call if needed there? ;)

**Comment by DavidZisky** [comment]
So... in our case we installed nginx on separate VM which act as proxy for two kyma clusters. On this VM we have SSL certificate for proxy URL and simple redirect nginx config. Then on both kyma clusters we did following to make it working properly:

1. edit "dex-config" configmap and add proxy VM URL into each "redirectURIs" section
2. by doing bit of a magic (overwriting nginx config in core-console deployment by adding new volume) we removed content-security-policy headers

So in order to make Kyma working properly behind proxy you would have to:

1. Allow user to specify proxy URL on installation (via env variable of whatever)
2. Put that URL together with kyma url into "dex-config" configmap
3. Adjust content-security-policy to respect proxy URL as well (we just removed the policy to make it easier but of course would be better to keep it and adjust)
